crates/stateless/src/witness_db.rs

```diff
@@ -0,0 +1,139 @@
+///
+/// This is designed for stateless execution scenarios where direct access to a full node's
+/// database is not available or desired.
+#[derive(Debug)]
+pub(crate) struct WitnessDatabase<'a> {
+    /// Map of block numbers to block hashes.
+    /// This is used to service the `BLOCKHASH` opcode.
+    // TODO: use Vec instead -- ancestors should be contiguous
+    // TODO: so we can use the current_block_number and an offset to
+    // TODO: get the block number of a particular ancestor
+    block_hashes_by_block_number: HashMap<u64, B256>,
+    /// Map of code hashes to bytecode.
+    /// Used to fetch contract code needed during execution.
+    bytecode: B256Map<Bytecode>,
+    /// The sparse Merkle Patricia Trie containing account and storage state.
+    /// This is used to provide account/storage values during EVM execution.
+    /// TODO: Ideally we do not have this trie and instead a simple map.
+    /// TODO: Then as a corollary we can avoid unnecessary hashing in `Database::storage`
+    /// TODO: and `Database::basic` without needing to cache the hashed Addresses and Keys
+    trie: &'a SparseStateTrie,
+}
+
+impl<'a> WitnessDatabase<'a> {
+    /// Creates a new [`WitnessDatabase`] instance.
+    ///
+    /// # Assumptions
+    ///
+    /// This function assumes:
+    /// 1. The provided `trie` has been populated with state data consistent with a known state root
+    ///    (e.g., using witness data and verifying against a parent block's state root).
+    /// 2. The `bytecode` map contains all bytecode corresponding to code hashes present in the
+    ///    account data within the `trie`.
+    /// 3. The `ancestor_hashes` map contains the block hashes for the relevant ancestor blocks (up
+    ///    to 256 including the current block number). It assumes these hashes correspond to a
+    ///    contiguous chain of blocks. The caller is responsible for verifying the contiguity and
+    ///    the block limit.
+    pub(crate) fn new(
+        trie: &'a SparseStateTrie,
+        bytecode: B256Map<Bytecode>,
+        ancestor_hashes: HashMap<u64, B256>,
+    ) -> Self {
+        Self { trie, block_hashes_by_block_number: ancestor_hashes, bytecode }
+    }
+}
+
+impl Database for WitnessDatabase<'_> {
+    /// The database error type.
+    type Error = ProviderError;
+
+    /// Get basic account information by hashing the address and looking up the account RLP
+    /// in the underlying [`SparseStateTrie`].
+    ///
+    /// Returns `Ok(None)` if the account is not found in the trie.
+    fn basic(&mut self, address: Address) -> Result<Option<AccountInfo>, Self::Error> {
+        let hashed_address = keccak256(address);
+        trace!(target: "reth-stateless::evm", %address, %hashed_address, "retrieving account");
+        let Some(bytes) = self.trie.get_account_value(&hashed_address) else {
+            trace!(target: "reth-stateless::evm", %address, %hashed_address, "no account found");
+            return Ok(None)
```

Comment:
here we should prove that the leaf is indeed not present in the trie. currently, this might lead to invalid execution due to partial witness